README.md
# lottie schema

基于 vscode 的 json schema https://lottie-lint.github.io/lottie-schema/animation.json

对 lottie json 进行解析

schema 关系 大图:https://www.processon.com/view/link/5c2ece6ae4b08a768398b06d

## how to run

```bash
npm install
# 打包
npm run build
# 发布 json schema
npm run deploy
```

## how to use

```js
// 初始化
const lottieSchema = new LottieSchema();
// 添加背景图片
lottieSchema.addBgImage({
  url: 'https://gw.alipayobjects.com/zos/antfincdn/I9Bu85KAaq/750x300.png',
  width: 750,
  height: 300,
});
// 获取lottie json
const animationData = lottieSchema.getObj()
```

```js
export default class LottieSchema {
    private lottieJSON;
    constructor(options?: any);
    getJSON(): string;
    getObj(): any;
    getVersion(): any;
    getSize(): {
        ip: any;
        op: any;
        width: any;
        height: any;
    };
    setSize({ width, height, }: {
        width: string;
        height: string;
    }): void;
    addBgImage({ url, width, height, }: {
        url: string;
        width: number;
        height: number;
    }): void;
    checkBgImageExist(): false | {
        bgAssetIdx: any;
        bgLayerIdx: any;
    };
    changeBgImage({ size, position, }: {
        size: any | undefined;
        position: any | undefined;
    }): void;
}

```


History.md

1.0.4 / 2019-08-13
==================

**features**
  * [[`f4b0bf5`](http://github.com/Lottie-Lint/lottie-schema/commit/f4b0bf5d397b0ab225a43f03f2bbf9d1c2b70794)] - feat: 添加默认options配置 (chenpx976 <<chenpx976@gmail.com>>)

**others**
  * [[`99b5d47`](http://github.com/Lottie-Lint/lottie-schema/commit/99b5d47c06f3f32462007f63d6d33354e028c398)] - Revert "feat: 去处 fonts 支持" (chenpx976 <<chenpx976@gmail.com>>)

1.0.3 / 2019-08-07
==================

**features**
  * [[`48397e8`](http://github.com/Lottie-Lint/lottie-schema/commit/48397e8ca9dd47ecba62202b52a5aecf84e2692c)] - feat: 去处 fonts 支持 (chenpx976 <<chenpx976@gmail.com>>)

1.0.2 / 2019-08-06
==================

**fixes**
  * [[`3460fd1`](http://github.com/Lottie-Lint/lottie-schema/commit/3460fd1143999ce05ee6c6bad9479d04fd16a78a)] - fix: 修复 ty 属性 (chenpx976 <<chenpx976@gmail.com>>)

1.0.1 / 2019-08-06
==================

**features**
  * [[`9e3743d`](http://github.com/Lottie-Lint/lottie-schema/commit/9e3743d3bcaf2ceeb3f3ca4eb86e4e06cdb75e6b)] - feat: 重构 (chenpx976 <<chenpx976@gmail.com>>)

1.0.0 / 2019-08-05
==================

**features**
  * [[`ec5b53f`](http://github.com/Lottie-Lint/lottie-schema/commit/ec5b53f384db2fae49866940968f9409d0a9e797)] - feat: 废除 immutable (chenpx976 <<chenpx976@gmail.com>>)
  * [[`2b49a4f`](http://github.com/Lottie-Lint/lottie-schema/commit/2b49a4f88235a7b3d5f149b2ce2f0b6972c22480)] - feat: 补充 文字 j (chenpx976 <<chenpx976@gmail.com>>)

0.0.13 / 2019-06-14
==================

**features**
  * [[`7b5ab1a`](http://github.com/Lottie-Lint/lottie-schema/commit/7b5ab1a5e69d608df4cb4edc148f9626fe41c337)] - feat: update v (chenpx976 <<chenpx976@gmail.com>>)
  * [[`f7d6e54`](http://github.com/Lottie-Lint/lottie-schema/commit/f7d6e5471a31e95200dcf22637d4ced0f122e2b3)] - feat: 添加动画播放规则schema (chenpx976 <<chenpx976@gmail.com>>)

0.0.12 / 2019-03-25
==================

**fixes**
  * [[`7499b72`](http://github.com/Lottie-Lint/lottie-schema/commit/7499b7226a7ad9fe621c1efec6d6b4f47cad3a5c)] - fix: 修复添加微动效的宽度计算问题 (chenpx976 <<chenpx976@gmail.com>>)

0.0.11 / 2019-03-25
==================

**fixes**
  * [[`a96c7ce`](http://github.com/Lottie-Lint/lottie-schema/commit/a96c7ce545b755fbfd6f22a126aafcebd9925cd7)] - fix: 修复添加微动效的属性计算问题 (chenpx976 <<chenpx976@gmail.com>>)

0.0.10 / 2019-03-20
==================

**features**
  * [[`2a401b8`](http://github.com/Lottie-Lint/lottie-schema/commit/2a401b87ab50399b9bd99834d1f4083bd7d56d52)] - feat: 添加修改图层逻辑 (chenpx976 <<chenpx976@gmail.com>>)

0.0.9 / 2019-03-15
==================

**features**
  * [[`548c51e`](http://github.com/Lottie-Lint/lottie-schema/commit/548c51e128e0bb3535e6840aa3e0a7f045a08890)] - feat: 去除id设置 (chenpx976 <<chenpx976@gmail.com>>)

**fixes**
  * [[`0204798`](http://github.com/Lottie-Lint/lottie-schema/commit/02047987b3e928bf7ad4d67d7a60f3d8d844c2e5)] - fix: 修复背景图层的判定 (chenpx976 <<chenpx976@gmail.com>>)

0.0.8 / 2019-03-13
==================

**features**
  * [[`718501d`](http://github.com/Lottie-Lint/lottie-schema/commit/718501db9db7b9813cb16c81d06fd4e82ee397ac)] - feat: 添加微动效 支持 字体 (chenpx976 <<chenpx976@gmail.com>>)
  * [[`74ee970`](http://github.com/Lottie-Lint/lottie-schema/commit/74ee9706f46db01f16fb16df9a79f898098bdc35)] - feat: update schema (chenpx976 <<chenpx976@gmail.com>>)
  * [[`64d9a19`](http://github.com/Lottie-Lint/lottie-schema/commit/64d9a1926b0791def3c8afd525411d0d81a7baca)] - feat: 更新 layer/test.json (chenpx976 <<chenpx976@gmail.com>>)

0.0.7 / 2019-03-08
==================

**features**
  * [[`802ac41`](http://github.com/Lottie-Lint/lottie-schema/commit/802ac41957cecfa934427a138ec79ff4d4027b27)] - feat: 解决 asset id 可能冲突的问题,为 precomp 确定layerId (chenpx976 <<chenpx976@gmail.com>>)
  * [[`0e40b5b`](http://github.com/Lottie-Lint/lottie-schema/commit/0e40b5b18ee410c7091cb244b0bb5e728c3faa84)] - feat: 添加schema中 image e 的描述 (chenpx976 <<chenpx976@gmail.com>>)

**fixes**
  * [[`f65b982`](http://github.com/Lottie-Lint/lottie-schema/commit/f65b9826c7a2c456b21734c02cc840bfeaa850d3)] - fix: 修复资源唯一id, 添加 资源的comp支持 (chenpx976 <<chenpx976@gmail.com>>)

0.0.6 / 2019-03-07
==================

**features**
  * [[`c2b9781`](http://github.com/Lottie-Lint/lottie-schema/commit/c2b9781664d42be593f96acc30fa7e89a35dd3a8)] - feat: 添加 precomp 功能 (chenpx976 <<chenpx976@gmail.com>>)
  * [[`8fd09eb`](http://github.com/Lottie-Lint/lottie-schema/commit/8fd09eb83f2c680f823aaf3edeac931bb862b877)] - feat: config -> options (chenpx976 <<chenpx976@gmail.com>>)
  * [[`5888a62`](http://github.com/Lottie-Lint/lottie-schema/commit/5888a62187b9be8b0de1aeb097b34c481bb65e7f)] - feat: update readme (chenpx976 <<chenpx976@gmail.com>>)

0.0.5 / 2019-03-07
==================

**features**
  * [[`f6ad73a`](http://github.com/Lottie-Lint/lottie-schema/commit/f6ad73aab2bce56a6a90ce8da0b48cfe0f61a38e)] - feat: update readme (chenpx976 <<chenpx976@gmail.com>>)

**fixes**
  * [[`a94bb6c`](http://github.com/Lottie-Lint/lottie-schema/commit/a94bb6c968dc52b861e43321a9902c22fc9446b8)] - fix: 修复背景图片适配问题 (chenpx976 <<chenpx976@gmail.com>>)

0.0.4 / 2019-03-06
==================

**features**
  * [[`949110d`](http://github.com/Lottie-Lint/lottie-schema/commit/949110dda15e9d7d387998036db9b7d8b959fbd5)] - feat: update readme (chenpx976 <<chenpx976@gmail.com>>)
  * [[`062896e`](http://github.com/Lottie-Lint/lottie-schema/commit/062896e812cde63d134b26e8ecb9d2292a0bfe62)] - feat: add format (chenpx976 <<chenpx976@gmail.com>>)
  * [[`d0440a4`](http://github.com/Lottie-Lint/lottie-schema/commit/d0440a4fd5c472cac6d87d6bf9254337590a94e3)] - feat: js -> ts (chenpx976 <<chenpx976@gmail.com>>)

0.0.3 / 2019-03-06
==================

**features**
  * [[`d0991df`](http://github.com/Lottie-Lint/lottie-schema/commit/d0991dff7992af9b90bcf03ab7dfa8fea8bf5cb4)] - feat: 去除new (chenpx976 <<chenpx976@gmail.com>>)
  * [[`42ed2d0`](http://github.com/Lottie-Lint/lottie-schema/commit/42ed2d0b7980f835b8e9dfd354f4cb70c18dd2d2)] - feat: Json -> JSON (chenpx976 <<chenpx976@gmail.com>>)

src/old.ts
// import { Map, List, fromJS } from 'immutable'

// // import LayerColor from './layers/color';
// import layerImage from './layers/image'
// import layerPrecomp from './layers/precomp'
// import assetImage from './assets/image'

// export default class LottieSchema {
//   private lottieJSON: any
//   public createLayerSize: number
//   constructor(options?: any) {
//     const defaultOptions = {
//       v: '5.5.5',
//       fr: 25,
//       ip: 0,
//       op: 750,
//       w: 1024,
//       h: 350,
//       nm: '初始化空层',
//       ddd: 0,
//     }
//     this.lottieJSON = Map({
//       assets: List(),
//       layers: List(),
//       markers: List(),
//       fonts: Map({
//         list: List(),
//       }),
//       ...defaultOptions,
//       ...options,
//     })
//     this.createLayerSize = 0
//   }

//   public getJSON() {
//     return JSON.stringify(this.lottieJSON)
//   }

//   public getObj() {
//     return this.lottieJSON.toJS()
//   }

//   public getVersion() {
//     return this.lottieJSON.get('v')
//   }

//   public getSize() {
//     return {
//       ip: this.lottieJSON.get('ip'),
//       op: this.lottieJSON.get('op'),
//       width: this.lottieJSON.get('w'),
//       height: this.lottieJSON.get('h'),
//     }
//   }

//   public setSize({ width, height }: { width: string; height: string }) {
//     if (width) {
//       this.lottieJSON = this.lottieJSON.set('w', width)
//     }
//     if (height) {
//       this.lottieJSON = this.lottieJSON.set('h', height)
//     }
//   }

//   /**
//    * getLayerProps
//    */
//   public getLayerProps() {
//     const layers = this.lottieJSON.get('layers')
//     return layers
//       .map(layer => ({
//         nm: layer.get('nm'),
//         w: layer.get('w'),
//         h: layer.get('h'),
//         s: layer.getIn(['ks', 's', 'k'])[0],
//         x: layer.getIn(['ks', 'p', 'k'])[0],
//         y: layer.getIn(['ks', 'p', 'k'])[1],
//       }))
//       .toJS()
//   }
//   /**
//    * delBgImage
//    */
//   public delBgImage() {
//     const imageIdx = this.checkBgImageExist()
//     if (imageIdx) {
//       this.createLayerSize -= 1
//       this.lottieJSON = this.lottieJSON
//         .set('assets', this.lottieJSON.get('assets').delete(imageIdx.assetIdx))
//         .set('layers', this.lottieJSON.get('layers').delete(imageIdx.layerIdx))
//     }
//   }
//   /**
//    * 两部分 assets 和 layers 通过 id 和 refId 关联起来, 讲 图片layer push 到 layers 最后
//    * 图片锚点: ks.a [ 图片宽度/2, 图片高度/2, 0]
//    * 图片缩放: ks.s [ 横向缩放, 纵向缩放, 100 ]
//    * 图片位移: ks.p [x, y, 100] 位移移动的位置是 图片锚点 相对于 画布的 x,y
//    * 要判断图片的大小和画布大小, 以 画布包含住 背景图片为目的, 实现 contain
//    */
//   public addBgImage({ url, width, height }: { url: string; width: number; height: number }) {
//     this.delBgImage()
//     // 获取画布属性
//     const { width: canvasWidth, height: canvasHeight, ip, op } = this.getSize()
//     const layers = this.lottieJSON.get('layers')
//     const assets = this.lottieJSON.get('assets')
//     // 设置背景图片宽高Url属性
//     const imageAsset = assetImage
//       .set('p', url)
//       .set('w', width)
//       .set('h', height)
//       .set('id', 'bgImage')
//     // 目标是 cover 包含适配, 以最大缩放为准
//     const wScale = Number.parseFloat(((canvasWidth / width) * 100).toFixed(3))
//     const hScale = Number.parseFloat(((canvasHeight / height) * 100).toFixed(3))
//     const scale = wScale < hScale ? hScale : wScale
//     this.createLayerSize += 1
//     const imageLayer = layerImage
//       .set('refId', 'bgImage')
//       .set('ind', this.createLayerSize)
//       // .set('ln', 'bgImage')
//       .set('cl', 'bgImage handlehook')
//       .set('nm', '背景图片')
//       .set('ip', ip)
//       .set('op', op)
//       .setIn(['ks', 'a', 'k'], [width / 2, height / 2, 0])
//       .setIn(['ks', 's', 'k'], [scale, scale, 0])
//       .setIn(['ks', 'p', 'k'], [canvasWidth / 2, canvasHeight / 2, 0])
//     this.lottieJSON = this.lottieJSON
//       .set('assets', assets.push(imageAsset))
//       .set('layers', layers.push(imageLayer))
//   }

//   public checkBgImageExist() {
//     const layers = this.lottieJSON.get('layers')
//     const assets = this.lottieJSON.get('assets')
//     const bgLayer = layers.last()
//     if (!bgLayer || !assets.size) {
//       return false
//     }
//     const assetIdx = assets.findIndex((value: any) => {
//       return value.get('id') === 'bgImage'
//     })
//     const layerIdx = layers.findIndex((value: any) => {
//       return value.get('refId') === 'bgImage'
//     })
//     if (assetIdx === -1 || layerIdx === -1) {
//       return false
//     }
//     if (!Number.isInteger(assetIdx) || !Number.isInteger(layerIdx)) {
//       return false
//     }
//     return {
//       assetIdx,
//       layerIdx,
//     }
//   }

//   public changeBgImage({ scale, position = {} }: { scale: number; position?: any }) {
//     // const { width, height } = size;
//     const { x, y } = position
//     const imageIdx = this.checkBgImageExist()
//     if (!imageIdx) {
//       return
//     }
//     // const { layerIdx } = imageIdx
//     this.changeLayersProp({ nm: '背景图片', scale, x, y })
//   }

//   /**
//    * getIdxByNm
//    */
//   public getIdxByNm(nm) {
//     return this.lottieJSON.get('layers').findIndex(value => value.get('nm') === nm)
//   }
//   /**
//    * changeLayersProp
//    */
//   public changeLayersProp({ nm, scale, x, y }) {
//     const layerIdx = this.getIdxByNm(nm)
//     if (layerIdx === -1) {
//       return
//     }
//     let layers = this.lottieJSON.get('layers').get(layerIdx)
//     // 图片缩放
//     if (scale) {
//       layers = layers.setIn(['ks', 's', 'k'], [scale, scale, 0])
//     }
//     // 图片位移
//     if (x || y) {
//       const [xx, yy] = layers.getIn(['ks', 'p', 'k'])
//       layers = layers.setIn(['ks', 'p', 'k'], [!x ? xx : x, !y ? yy : y, 0])
//     }
//     // 图片资源替换的情况,在外围调用
//     this.lottieJSON = this.lottieJSON.set(
//       'layers',
//       this.lottieJSON.get('layers').set(layerIdx, layers)
//     )
//   }

//   public isLottieJSON(jsonObj) {
//     if (
//       typeof jsonObj.fr === 'number' &&
//       typeof jsonObj.ip === 'number' &&
//       typeof jsonObj.op === 'number' &&
//       typeof jsonObj.w === 'number' &&
//       typeof jsonObj.h === 'number' &&
//       typeof jsonObj.v === 'string' &&
//       Array.isArray(jsonObj.assets) &&
//       Array.isArray(jsonObj.layers)
//     ) {
//       return true
//     }

//     return false
//   }

//   /**
//    * addPrecomp 合并 lottie 到 当前lottie文件当中
//    * 入参 要合并的obj 做fromJS处理
//    * 取出 fr ip op w h nm , layers
//    * 与 this.lottieJSON 的 基础属性做对比 一致的情况下进行下一步
//    * 新建 asset 空Map: precompAsset { id, layers}  确定一个随机id 并塞入 layers
//    * 新建 layer: layerPrecomp 设置 refId
//    * 将 asset layer unshift 进入 asserts 和 layers
//    */
//   public addPrecomp(jsonObj) {
//     // 检测 jsonObj 合法性
//     if (!this.isLottieJSON(jsonObj)) {
//       return
//     }
//     // 加一个 assets id 的处理, 添加一个后缀
//     const time = new Date().valueOf()
//     const timeSuffix = time + '_'
//     const timePrefix = '_' + time
//     const layerId = 'microLottie' + timePrefix
//     const formatJSON = JSON.stringify(jsonObj)
//       .replace(/"id":"image_/g, '"id":"image_' + timeSuffix)
//       .replace(/"refId":"image_/g, '"refId":"image_' + timeSuffix)
//       .replace(/"id":"comp_/g, '"id":"comp_' + timeSuffix)
//       .replace(/"refId":"comp_/g, '"refId":"comp_' + timeSuffix)
//     const precomp = fromJS(JSON.parse(formatJSON))
//     // 检查基础属性是否一致
//     const pfr = precomp.get('fr')
//     const pip = precomp.get('ip')
//     const pop = precomp.get('op')
//     const width = precomp.get('w')
//     const height = precomp.get('h')
//     const fr = this.lottieJSON.get('fr')
//     const ip = this.lottieJSON.get('ip')
//     const op = this.lottieJSON.get('op')
//     const canvasWidth = this.lottieJSON.get('w')
//     const canvasHeight = this.lottieJSON.get('h')
//     if (!fr === pfr && op === pop && ip === pip) {
//       return
//     }
//     // 将 layer 转换 为 asset
//     const asset = fromJS({
//       id: layerId,
//       layers: precomp.get('layers'),
//     })
//     // 迁移 微动效 中的 assets
//     const assets = precomp.get('assets').push(asset)
//     // 迁移 微动效 中的 fonts
//     const fontsList = precomp.hasIn(['fonts', 'list']) ? precomp.getIn(['fonts', 'list']) : []
//     // 目标是 contain 包含适配, 以最小缩放为准
//     const wScale = Number.parseFloat(((canvasWidth / width) * 100).toFixed(3))
//     const hScale = Number.parseFloat(((canvasHeight / height) * 100).toFixed(3))
//     const scale = wScale > hScale ? hScale : wScale
//     this.createLayerSize += 1
//     const layer = layerPrecomp
//       .set('refId', layerId)
//       .set('ind', this.createLayerSize)
//       .set('nm', 'precomp' + timePrefix)
//       .set('ln', 'precomp' + timePrefix)
//       .set('w', Number(width))
//       .set('h', Number(height))
//       .set('cl', 'precomp handlehook')
//       .setIn(['ks', 'a', 'k'], [width / 2, height / 2, 0]) // 锚点 是相对于本身的
//       .setIn(['ks', 's', 'k'], [scale, scale, 0])
//       .setIn(['ks', 'p', 'k'], [canvasWidth / 2, canvasHeight / 2, 0]) // 位置是相对于画布的

//     this.lottieJSON = this.lottieJSON
//       .set('assets', this.lottieJSON.get('assets').merge(assets))
//       .set('layers', this.lottieJSON.get('layers').unshift(layer))
//       .setIn(['fonts', 'list'], this.lottieJSON.getIn(['fonts', 'list']).merge(fontsList))
//   }

//   /**
//    * delPrecomp 删除微动效
//    * 查找 assets 里面 refId
//    */
//   public delPrecomp(nm: string) {
//     const layerIdx = this.getIdxByNm(nm)
//     const layerId = 'microLottie' + '_' + nm.split('_')[1]
//     const assetIdx = this.lottieJSON.get('assets').findIndex(value => value.id === layerId)
//     if (assetIdx > -1) {
//       this.lottieJSON = this.lottieJSON.set(
//         'assets',
//         this.lottieJSON.get('assets').delete(assetIdx)
//       )
//     }
//     if (layerIdx > -1) {
//       this.lottieJSON = this.lottieJSON.set(
//         'layers',
//         this.lottieJSON.get('layers').delete(layerIdx)
//       )
//       this.createLayerSize -= 1
//     }
//   }
//   /* addBgColor(color) {
//     const {
//       width,
//       height,
//       ip,
//       op,
//     } = this.getSize();
//     const layers = this.lottieJSON.get('layers')
//     const colorLayer = LayerColor.set('sc', color)
//       .set('nm', '背景颜色:' + color)
//       .set('sw', width)
//       .set('sh', height)
//       .set('ip', ip)
//       .set('op', op)
//       .set('cl', 'handlehook')
//       .set('ln', 'bgColor')
//       .setIn(['ks', 'p', 'k'], [width / 2, height / 2, 0])
//       .setIn(['ks', 'a', 'k'], [width / 2, height / 2, 0])
//     this.lottieJSON = this.lottieJSON.set('layers', layers.push(colorLayer));
//   }

//   delBgLayer() {
//     this.lottieJSON = this.lottieJSON.set('layers', this.lottieJSON.get('layers').pop())
//   } */
// }
